packages/@ember/-internals/views/lib/system/event_dispatcher.ts

```diff
@@ -239,50 +249,58 @@ export default EmberObject.extend({
     @param {String} event the name of the event in the browser
     @param {String} eventName the name of the method to call on the view
   */
-  setupHandler(rootElement, event, eventName) {
+  setupHandler(rootElement: Element, event: string, eventName: string | null) {
     if (eventName === null || !this.lazyEvents.has(event)) {
       return; // nothing to do
     }
 
-    let viewHandler = (target, event) => {
-      let view = getElementView(target);
+    let viewHandler = (target: Element, event: Event) => {
+      // SAFETY: SimpleElement is supposed to be a subset of Element so this _should_ be safe.
+      // However, the types are more specific in some places which necessitates the `as`.
+      let view = getElementView(target as unknown as SimpleElement);
       let result = true;
 
       if (view) {
-        result = view.handleEvent(eventName, event);
+        // SAFETY: As currently written, this is not safe. Though it seems to always be true.
+        result = (view as Component).handleEvent(eventName, event);
       }
 
       return result;
     };
 
-    let actionHandler = (target, event) => {
+    let actionHandler = (target: Element, event: Event) => {
       let actionId = target.getAttribute('data-ember-action');
-      let actions = ActionManager.registeredActions[actionId];
+      let actions: Array<ActionState> | undefined;
 
       // In Glimmer2 this attribute is set to an empty string and an additional
       // attribute it set for each action on a given element. In this case, the
       // attributes need to be read so that a proper set of action handlers can
       // be coalesced.
       if (actionId === '') {
-        let attributes = target.attributes;
-        let attributeCount = attributes.length;
-
         actions = [];
 
-        for (let i = 0; i < attributeCount; i++) {
-          let attr = attributes.item(i);
+        for (let attr of target.attributes) {
           let attrName = attr.name;
 
           if (attrName.indexOf('data-ember-action-') === 0) {
-            actions = actions.concat(ActionManager.registeredActions[attr.value]);
+            let action = ActionManager.registeredActions[attr.value];
+            assert('[BUG] Missing action', action);
+            actions.push(action);
           }
         }
+      } else if (actionId) {
+        // FIXME: This branch is never called in tests. Improve tests or remove
+        let actionState = ActionManager.registeredActions[actionId];
+        if (actionState) {
+          actions = [actionState];
+        }
       }
 
       // We have to check for actions here since in some cases, jQuery will trigger
       // an event on `removeChild` (i.e. focusout) after we've already torn down the
       // action handlers for the view.
       if (!actions) {
+        // FIXME: This branch is never called in tests. Improve tests or remove
```

Comment:
Not sure if there's anything more we should do about this right now.

Reply:
Huh. Is it even an issue anymore given that we don't have jQuery *integration* anymore?

Reply:
I suspect we just don't need it. But am also a bit hesitant to remove it without fully understanding.